File: lib/ch2/2.2.js
// Implement an algorithm to find the nth to last element of a singly
// linked list.

const nthToLast = (list, n) => {
  let node1 = list;
  let node2 = list;
  for (let i = 0; i < n; i++) {
    if (!node2.next) return null;
    node2 = node2.next;
  }

  while (node2.next) {
    node1 = node1.next;
    node2 = node2.next;
  }

  return node1;
};

module.exports = nthToLast;

File: java/ch2/2.2.java
public Node nthToLast(Node head, int n) {
    if (head == null) return head;
    Node slow = head;
    Node fast = head;
    for (int i = 0; i < n; i++) {
        if (fast.next == null) return null;
        fast = fast.next;
    }
    while (fast.next != null) {
        slow = slow.next;
        fast = fast.next;
    }
    return slow;
}
